### PR TITLE
Fix duration=0 & stream disabled bugs

### PR DESCRIPTION
**Pull request by hmstepanek**
# Overview
Fix the following bugs in openai:
* During errors, the function trace is not exited before capturing the duration so the duration=0 inside the events.
* When not streaming but stream monitoring is disabled, the function trace is exited early and events are not recorded.

### CODE REVIEW

File: newrelic/hooks/mlmodel_openai.py [hunk 1]
```diff
@@ -367,13 +373,13 @@ def _handle_completion_success(transaction, linking_metadata, completion_id, kwa
     trace_id = linking_metadata.get("trace.id")
     request_message_list = kwargs.get("messages", [])
     stream = kwargs.get("stream", False)
-    # If response is not a stream generator or stream monitoring is disabled, we exit
-    # the function trace now.
-    if not stream or not settings.ai_monitoring.streaming.enabled:
+    # Only if streaming and streaming monitoring is enabled and the response is not empty
+    # do we not exit the function trace.
+    if not (stream and settings.ai_monitoring.streaming.enabled and return_val):
```
Comment (hmstepanek) [inline newrelic/hooks/mlmodel_openai.py:378]:
```suggestion
    if not stream or not settings.ai_monitoring.streaming.enabled or not return_val:
```

File: tests/mlmodel_openai/test_chat_completion.py [hunk 2]
```diff
@@ -283,6 +284,25 @@ def test_openai_chat_completion_sync_in_txn_no_llm_metadata(set_trace_info):
...
+# One summary event, one system message, one user message, and one response message from the assistant
+@validate_custom_event_count(count=4)
+@validate_transaction_metrics(
+    "test_chat_completion:test_openai_chat_completion_sync_in_txn_no_llm_metadata_stream_monitoring_disabled",
+    scoped_metrics=[("Llm/completion/OpenAI/create", 1)],
+    rollup_metrics=[("Llm/completion/OpenAI/create", 1)],
+    background_task=True,
+)
+@background_task()
+def test_openai_chat_completion_sync_in_txn_no_llm_metadata_stream_monitoring_disabled(set_trace_info):
```
Comment (umaannamalai) [inline tests/mlmodel_openai/test_chat_completion.py:299]: I realize some of our test names are getting really long and the names are implying we are trying to test multiple things in a single test. Given we aren't testing a case where LLM metadata is present when streaming is disabled, we could probably drop the "`no_llm_metadata`" part in the test name.

File: tests/mlmodel_openai/test_chat_completion.py [hunk 3]
```diff
@@ -311,6 +312,25 @@ def test_openai_chat_completion_sync_in_txn_no_llm_metadata(set_trace_info):
...
```
Comment (hmstepanek) [inline tests/mlmodel_openai/test_chat_completion.py:327]:
```suggestion
def test_openai_chat_completion_sync_in_txn_stream_monitoring_disabled(set_trace_info):
```

File: tests/mlmodel_openai/test_chat_completion.py [hunk 4]
```diff
@@ -350,6 +370,27 @@ def test_openai_chat_completion_async_no_llm_metadata(loop, set_trace_info):
...
+@validate_custom_events(chat_completion_recorded_events_no_llm_metadata)
+@validate_custom_event_count(count=4)
+@validate_transaction_metrics(
+    "test_chat_completion:test_openai_chat_completion_async_no_llm_metadata_stream_monitoring_disabled",
+    scoped_metrics=[("Llm/completion/OpenAI/acreate", 1)],
+    rollup_metrics=[("Llm/completion/OpenAI/acreate", 1)],
+    background_task=True,
+)
+@background_task()
+def test_openai_chat_completion_async_no_llm_metadata_stream_monitoring_disabled(loop, set_trace_info):
```
Comment (hmstepanek) [inline tests/mlmodel_openai/test_chat_completion.py:384]:
```suggestion
def test_openai_chat_completion_async_stream_monitoring_disabled(loop, set_trace_info):
```

File: tests/mlmodel_openai/test_chat_completion_v1.py [hunk 5]
```diff
@@ -306,6 +306,27 @@ def test_openai_chat_completion_sync_in_txn_no_llm_metadata(set_trace_info, sync
...
+# One summary event, one system message, one user message, and one response message from the assistant
+@validate_custom_event_count(count=4)
+@validate_transaction_metrics(
+    "test_chat_completion_v1:test_openai_chat_completion_sync_in_txn_no_llm_metadata_stream_monitoring_disabled",
+    scoped_metrics=[("Llm/completion/OpenAI/create", 1)],
+    rollup_metrics=[("Llm/completion/OpenAI/create", 1)],
+    background_task=True,
+)
+@background_task()
+def test_openai_chat_completion_sync_in_txn_no_llm_metadata_stream_monitoring_disabled(
```
Comment (hmstepanek) [inline tests/mlmodel_openai/test_chat_completion_v1.py:321]:
```suggestion
def test_openai_chat_completion_sync_in_txn_stream_monitoring_disabled(
```

File: tests/mlmodel_openai/test_chat_completion_v1.py [hunk 6]
```diff
@@ -346,6 +367,29 @@ def test_openai_chat_completion_async_no_llm_metadata(loop, set_trace_info, asyn
...
+@validate_custom_events(chat_completion_recorded_events_no_llm_metadata)
+@validate_custom_event_count(count=4)
+@validate_transaction_metrics(
+    "test_chat_completion_v1:test_openai_chat_completion_async_no_llm_metadata_stream_monitoring_disabled",
+    scoped_metrics=[("Llm/completion/OpenAI/create", 1)],
+    rollup_metrics=[("Llm/completion/OpenAI/create", 1)],
+    background_task=True,
+)
+@background_task()
+def test_openai_chat_completion_async_no_llm_metadata_stream_monitoring_disabled(
```
Comment (hmstepanek) [inline tests/mlmodel_openai/test_chat_completion_v1.py:381]:
```suggestion
def test_openai_chat_completion_async_stream_monitoring_disabled(
```

File: tests/mlmodel_openai/test_chat_completion.py [hunk 7]
```diff
@@ -311,6 +312,25 @@ def test_openai_chat_completion_sync_in_txn_no_llm_metadata(set_trace_info):
     )
 
 
+@reset_core_stats_engine()
+@disabled_ai_monitoring_streaming_settings
+@validate_custom_events(chat_completion_recorded_events_no_llm_metadata)
+# One summary event, one system message, one user message, and one response message from the assistant
+@validate_custom_event_count(count=4)
+@validate_transaction_metrics(
+    "test_chat_completion:test_openai_chat_completion_sync_in_txn_no_llm_metadata_stream_monitoring_disabled",
```
Comment (hmstepanek) [inline tests/mlmodel_openai/test_chat_completion.py:321]:
```suggestion
    "test_chat_completion:test_openai_chat_completion_sync_in_txn_stream_monitoring_disabled",
```

File: tests/mlmodel_openai/test_chat_completion.py [hunk 8]
```diff
@@ -350,6 +370,27 @@ def test_openai_chat_completion_async_no_llm_metadata(loop, set_trace_info):
     )
 
 
+@reset_core_stats_engine()
+@disabled_ai_monitoring_streaming_settings
+@validate_custom_events(chat_completion_recorded_events_no_llm_metadata)
+@validate_custom_event_count(count=4)
+@validate_transaction_metrics(
+    "test_chat_completion:test_openai_chat_completion_async_no_llm_metadata_stream_monitoring_disabled",
```
Comment (hmstepanek) [inline tests/mlmodel_openai/test_chat_completion.py:378]:
```suggestion
    "test_chat_completion:test_openai_chat_completion_async_stream_monitoring_disabled",
```

File: tests/mlmodel_openai/test_chat_completion_v1.py [hunk 9]
```diff
@@ -306,6 +306,27 @@ def test_openai_chat_completion_sync_in_txn_no_llm_metadata(set_trace_info, sync
     )
 
 
+@reset_core_stats_engine()
+@disabled_ai_monitoring_streaming_settings
+@validate_custom_events(chat_completion_recorded_events_no_llm_metadata)
+# One summary event, one system message, one user message, and one response message from the assistant
+@validate_custom_event_count(count=4)
+@validate_transaction_metrics(
+    "test_chat_completion_v1:test_openai_chat_completion_sync_in_txn_no_llm_metadata_stream_monitoring_disabled",
```
Comment (hmstepanek) [inline tests/mlmodel_openai/test_chat_completion_v1.py:315]:
```suggestion
    "test_chat_completion_v1:test_openai_chat_completion_sync_in_txn_stream_monitoring_disabled",
```

File: tests/mlmodel_openai/test_chat_completion_v1.py [hunk 10]
```diff
@@ -346,6 +367,29 @@ def test_openai_chat_completion_async_no_llm_metadata(loop, set_trace_info, asyn
     )
 
 
+@reset_core_stats_engine()
+@disabled_ai_monitoring_streaming_settings
+@validate_custom_events(chat_completion_recorded_events_no_llm_metadata)
+@validate_custom_event_count(count=4)
+@validate_transaction_metrics(
+    "test_chat_completion_v1:test_openai_chat_completion_async_no_llm_metadata_stream_monitoring_disabled",
```
Comment (hmstepanek) [inline tests/mlmodel_openai/test_chat_completion_v1.py:375]:
```suggestion
    "test_chat_completion_v1:test_openai_chat_completion_async_stream_monitoring_disabled",
```

**Review by umaannamalai — APPROVED** [review]
This looks good - testing for a non-zero duration is going to add far more complexity than it is worth.

**Pull request merged**